File: loop_engineering/checklists/key_diffs_gate.py
"""key-diffs.yaml 硬 gate (design §2.3).

§2.3: risk:high 或 exclusive:true 的 task, 收口前 key-diffs.yaml 必须
存在、可解析、且 key_diffs 非空 —— 机制硬 gate. 其它 task 是软约束
(文件可选, 不强制).

调用方 (收口阶段) 用 validate_many + all_hard_gates_pass 检验:
任一硬 gate task FAIL 则整体不能进 COMPLETE.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from loop_engineering.schema.artifacts import KeyDiffsFile
from loop_engineering.schema.task_plan import RiskLevel, Task


class GateStatus(StrEnum):
    """gate 校验结果三态."""

    PASS = "pass"  # 硬 gate 通过 / 普通 task 自愿提交且非空
    FAIL = "fail"  # 硬 gate 失败 (缺文件 / 空 / 解析失败)
    SOFT = "soft"  # 软约束未满足 (普通 task 缺 key-diffs.yaml), 不阻断


@dataclass(frozen=True)
class KeyDiffsGateResult:
    """单个 task 的 key-diffs gate 校验结果."""

    task_id: str
    status: GateStatus
    reason: str


def is_hard_gate_task(task: Task) -> bool:
    """该 task 是否触发 key-diffs 硬 gate.

    design §2.3: risk==high 或 exclusive==True -> 硬 gate.
    """
    return task.risk == RiskLevel.high or bool(task.exclusive)


def validate_key_diffs_submission(
    task: Task,
    key_diffs: KeyDiffsFile | None,
    *,
    raw_yaml_text: str | None = None,
) -> KeyDiffsGateResult:
    """校验单个 task 的 key-diffs 提交.

    Args:
        task: 计划中的 task.
        key_diffs: 已解析的 KeyDiffsFile (若 None 表示文件缺失或解析失败).
        raw_yaml_text: 调用方可传入原始 YAML 文本用于诊断 (例如解析失败时回显).
            本函数本身不重新解析 YAML (调用方负责 try/except); 此参数仅用于
            错误信息富化, 可省略.

    Returns:
        KeyDiffsGateResult: task_id + status + reason.
    """
    hard = is_hard_gate_task(task)

    if hard:
        if key_diffs is None:
            tail = f" (raw_yaml_text 前 80 字符: {raw_yaml_text[:80]!r})" if raw_yaml_text else ""
            return KeyDiffsGateResult(
                task_id=task.id,
                status=GateStatus.FAIL,
                reason=f"硬 gate task 缺 key-diffs.yaml 或解析失败{tail}",
            )
        if not key_diffs.is_meaningful():
            return KeyDiffsGateResult(
                task_id=task.id,
                status=GateStatus.FAIL,
                reason="硬 gate task 的 key_diffs 为空 (must be 非空)",
            )
        return KeyDiffsGateResult(
            task_id=task.id,
            status=GateStatus.PASS,
            reason=f"硬 gate 通过: {len(key_diffs.key_diffs)} 条 key diff",
        )

    # 普通 task —— 软约束
    if key_diffs is None or not key_diffs.is_meaningful():
        return KeyDiffsGateResult(
            task_id=task.id,
            status=GateStatus.SOFT,
            reason="普通 task, key-diffs.yaml 可省 (软约束)",
        )
    return KeyDiffsGateResult(
        task_id=task.id,
        status=GateStatus.PASS,
        reason=f"普通 task 自愿提交: {len(key_diffs.key_diffs)} 条 key diff",
    )


def validate_many(
    tasks: list[Task],
    key_diffs_by_task: dict[str, KeyDiffsFile | None],
) -> list[KeyDiffsGateResult]:
    """批量校验: 每 task 一条结果.

    Args:
        tasks: 计划中的全部 task.
        key_diffs_by_task: task_id -> KeyDiffsFile | None.

    Returns:
        与 tasks 一一对应的 KeyDiffsGateResult 列表.
    """
    results: list[KeyDiffsGateResult] = []
    for t in tasks:
        kd = key_diffs_by_task.get(t.id)
        results.append(validate_key_diffs_submission(t, kd))
    return results


def all_hard_gates_pass(results: list[KeyDiffsGateResult]) -> bool:
    """是否所有硬 gate task 都 PASS.

    用于收口阶段: 任一 FAIL status 的硬 gate task 存在 -> 不能进 COMPLETE.
    SOFT 状态 (普通 task 缺文件) 不阻断.
    """
    return all(r.status != GateStatus.FAIL for r in results)

File: packages/ssot-ts/src/checklists/key_diffs_gate.ts
/**
 * key-diffs.yaml 硬 gate (design §2.3) —— TS 版, 等价 Python
 * `loop_engineering/checklists/key_diffs_gate.py`。
 *
 * §2.3: risk:high 或 exclusive:true 的 task, 收口前 key-diffs.yaml 必须
 * 存在、可解析、且 key_diffs 非空 —— 机制硬 gate。其它 task 是软约束
 * (文件可选, 不强制)。
 *
 * 调用方 (收口阶段) 用 validateMany + allHardGatesPass 检验:
 * 任一硬 gate task FAIL 则整体不能进 COMPLETE。
 *
 * 与 Python 的差异处理:
 * - StrEnum GateStatus → const 对象 + 字符串字面量联合类型 (值即字符串, 等价)。
 * - dataclass(frozen=True) → readonly 接口。
 * - KeyDiffsFile.is_meaningful() 实例方法 → schema 子包的独立函数 isMeaningful(file)。
 * - f"{x!r}" repr → pyRepr (字符串加单引号), 测试只断言关键子串。
 */
import { isMeaningful } from "../schema/artifacts.js";
import type { KeyDiffsFile } from "../schema/artifacts.js";
import { RiskLevel } from "../schema/task_plan.js";
import type { Task } from "../schema/task_plan.js";

/** gate 校验结果三态 (StrEnum 等价: 值即字符串)。 */
export const GateStatus = {
  PASS: "pass", // 硬 gate 通过 / 普通 task 自愿提交且非空
  FAIL: "fail", // 硬 gate 失败 (缺文件 / 空 / 解析失败)
  SOFT: "soft", // 软约束未满足 (普通 task 缺 key-diffs.yaml), 不阻断
} as const;
export type GateStatus = (typeof GateStatus)[keyof typeof GateStatus];

/** 单个 task 的 key-diffs gate 校验结果。 */
export interface KeyDiffsGateResult {
  readonly task_id: string;
  readonly status: GateStatus;
  readonly reason: string;
}

/** 把值渲染成近似 Python `repr` (字符串加单引号), 用于诊断信息回显。 */
function pyRepr(v: unknown): string {
  if (typeof v === "string") {
    return `'${v}'`;
  }
  return String(v);
}

/**
 * 该 task 是否触发 key-diffs 硬 gate。
 *
 * design §2.3: risk==high 或 exclusive==True -> 硬 gate。
 */
export function isHardGateTask(task: Task): boolean {
  return task.risk === RiskLevel.high || Boolean(task.exclusive);
}

/**
 * 校验单个 task 的 key-diffs 提交。
 *
 * @param task 计划中的 task。
 * @param keyDiffs 已解析的 KeyDiffsFile (若 null/undefined 表示文件缺失或解析失败)。
 * @param rawYamlText 调用方可传入原始 YAML 文本用于诊断 (例如解析失败时回显)。
 *   本函数本身不重新解析 YAML (调用方负责 try/catch); 此参数仅用于错误信息富化, 可省略。
 */
export function validateKeyDiffsSubmission(
  task: Task,
  keyDiffs: KeyDiffsFile | null | undefined,
  options?: { rawYamlText?: string | null },
): KeyDiffsGateResult {
  const rawYamlText = options?.rawYamlText ?? null;
  const hard = isHardGateTask(task);

  if (hard) {
    if (keyDiffs === null || keyDiffs === undefined) {
      const tail = rawYamlText
        ? ` (raw_yaml_text 前 80 字符: ${pyRepr(rawYamlText.slice(0, 80))})`
        : "";
      return {
        task_id: task.id,
        status: GateStatus.FAIL,
        reason: `硬 gate task 缺 key-diffs.yaml 或解析失败${tail}`,
      };
    }
    if (!isMeaningful(keyDiffs)) {
      return {
        task_id: task.id,
        status: GateStatus.FAIL,
        reason: "硬 gate task 的 key_diffs 为空 (must be 非空)",
      };
    }
    return {
      task_id: task.id,
      status: GateStatus.PASS,
      reason: `硬 gate 通过: ${keyDiffs.key_diffs.length} 条 key diff`,
    };
  }

  // 普通 task —— 软约束
  if (keyDiffs === null || keyDiffs === undefined || !isMeaningful(keyDiffs)) {
    return {
      task_id: task.id,
      status: GateStatus.SOFT,
      reason: "普通 task, key-diffs.yaml 可省 (软约束)",
    };
  }
  return {
    task_id: task.id,
    status: GateStatus.PASS,
    reason: `普通 task 自愿提交: ${keyDiffs.key_diffs.length} 条 key diff`,
  };
}

/**
 * 批量校验: 每 task 一条结果。
 *
 * @param tasks 计划中的全部 task。
 * @param keyDiffsByTask task_id -> KeyDiffsFile | null。
 * @returns 与 tasks 一一对应的 KeyDiffsGateResult 列表。
 */
export function validateMany(
  tasks: readonly Task[],
  keyDiffsByTask: Record<string, KeyDiffsFile | null | undefined> | Map<string, KeyDiffsFile | null | undefined>,
): KeyDiffsGateResult[] {
  const getKd = (id: string): KeyDiffsFile | null | undefined =>
    keyDiffsByTask instanceof Map ? keyDiffsByTask.get(id) : keyDiffsByTask[id];

  const results: KeyDiffsGateResult[] = [];
  for (const t of tasks) {
    results.push(validateKeyDiffsSubmission(t, getKd(t.id)));
  }
  return results;
}

/**
 * 是否所有硬 gate task 都 PASS。
 *
 * 用于收口阶段: 任一 FAIL status 的硬 gate task 存在 -> 不能进 COMPLETE。
 * SOFT 状态 (普通 task 缺文件) 不阻断。
 */
export function allHardGatesPass(results: readonly KeyDiffsGateResult[]): boolean {
  return results.every((r) => r.status !== GateStatus.FAIL);
}
